File: Homework 6/Sources/GTUSet.cpp
#include "../Headers/GTUSet.h"
namespace GTU
{
    template <class T>
    GTUSet<T>::GTUSet(int size) : GTUSetBase<T>(), _capacity(size), _size(0)
    {
        if(size < 0) throw std::bad_pafram();
        _set = std::shared_ptr<T>(new T [_capacity], std::default_delete<T[]>());
    }

    template <class T>
    GTUSet<T>::GTUSet(const GTUSetBase<T> & other) : GTUSet(other.max_size())
    {
        *this = other;
    }

    template <class T>
    GTUSet<T>::GTUSet(const typename GTUSetBase<T>::GTUIterator & first, const typename GTUSetBase<T>::GTUIterator & last) : GTUSet()
    {
        for(typename GTUSetBase<T>::GTUIterator temp = first; temp != last; ++temp)
            insert(*temp);
    }

    template <class T>
    const GTUSet<T> & GTUSet<T>::operator = (const GTUSetBase<T> & other)
    {
        if(max_size() != other.max_size())
        {
            _set = nullptr;
            _capacity = other.max_size();
            _set = std::shared_ptr<T>(new T [_capacity], std::default_delete<T[]>());
        }
        _size = 0;
        for(typename GTUSetBase<T>::GTUIterator counter = other.begin(); counter != other.end(); ++counter)
        {
            *(_set.get() + _size) = *(counter);
        }
        return *(this);
    }
    
    template <class T>
    bool GTUSet<T>::empty() const noexcept
    {
        return (_size == 0);
    }
    
    template <class T>
    int GTUSet<T>::size() const noexcept
    {
        return _size;
    }
    
    template <class T>
    int GTUSet<T>::max_size() const noexcept
    {
        return _capacity;
    }
    
    template <class T>
    void GTUSet<T>::insert(const T & data) throw(std::bad_pafram)
    {
        if(count(data) != 0)
            throw new std::bad_pafram();
        if(_size == _capacity)
        {
            _capacity *= 2;
            std::shared_ptr<T> temp(_set);
            _set = std::shared_ptr<T>(new T [_capacity], std::default_delete<T[]>());
            for(int i = 0; i < _size; ++i)
                *(_set.get() + i) = *(temp.get() + i);
        }
        *(_set.get() + _size++) = data;
    }

    template <class T>
    void GTUSet<T>::erase(const T & data)
    {
        if(count(data) == 0)
            throw std::bad_pafram();
        for(int i = 0; i < _size; ++i)
            if(*(_set.get() + i) == data)
            {
                while(i < _size)
                {
                    *(_set.get() + i) = *(_set.get() + (i + 1));
                    ++i;
                }
            }
    }
    
    template <class T>
    void GTUSet<T>::clear() noexcept
    {
        *this = GTUSet<T>();
    }

    template <class T>
    typename GTUSetBase<T>::GTUIterator GTUSet<T>::find(const T & data) const noexcept
    {
        int i;
        for(i = 0; *(_set.get() + i) != data && i < _size; ++i);
        return _set.get() + i;
    }

    template <class T>
    int GTUSet<T>::count(const T & data) const noexcept 
    {
        int temp_c = 0;
        for(int i = 0; i < _size; ++i)
            if(*(_set.get() + i) == data)
                ++temp_c;
        return temp_c;
    }

    template <class T>
    typename GTUSetBase<T>::GTUIterator GTUSet<T>::begin() const noexcept
    {
        return _set;
    }
    
    template <class T>
    typename GTUSetBase<T>::GTUIterator GTUSet<T>::end() const noexcept
    {
        return _set.get() + _size;
    }
}

File: Homework 7/151044062/151044062/151044062/src/GTUSet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

import java.security.InvalidParameterException;

/**
 *
 * @author Muaz Kurt, 151044062
 * @param <T>
 */
public class GTUSet<T> implements GTUSetInt<T>
{
    private Object[] _set;
    private int _size;
    private int _capacity;
    public GTUSet()
    {
        _size = 0;
        _capacity = 10;
        _set = (T[]) new Object [_capacity];
    }
    
    public GTUSet(int input)
    {
        _size = 0;
        if(input > 0)
            _capacity = input;
        else
            _capacity = 10;
        _set = (T[]) new Object [_capacity];
    }

    public GTUSet(GTUSetInt<T> other)
    {
        _size = 0;
        _capacity = other.max_size();
        _set = (T[]) new Object [_capacity];
        for(GTUIterator<T> friend = other.begin(); friend.hasNext(); friend.next())
            insert(friend.index());
    }

    public GTUSet(GTUIterator<T> start, GTUIterator<T> finish)
    {
        _size = 0;
        _capacity = 10;
        _set = (T[]) new Object [_capacity];
        for(GTUIterator<T> friend = start; friend.hasNext();)
            insert(friend.next());
    }

    @Override
    public boolean empty()
    {
        return !(_size > 0);
    }
    @Override
    public int size()
    {
        return _size;
    }
    @Override
    public int max_size()
    {
        return _capacity;
    }
    @Override
    public void insert(Object input) throws InvalidParameterException
    {
        if(count(input) > 0)
            throw new InvalidParameterException();
        if(_size == _capacity - 1)
        {
            T [] temp = (T[])_set;
            _capacity *= 2;
            _set = (T[]) new Object [_capacity];
            for(int i = 0; i < _size; ++i)
                _set[i] = temp[i];
        }
        _set[_size++] = input;
    }
    @Override
    public GTUSetInt<T> intersection(GTUSetInt<T> other)
    {
        GTUSet<T> intersect = new GTUSet<>();
        for(GTUIterator<T> my = begin(); my.hasNext(); my.next())
            if(other.count(my.index()) == 1)
                intersect.insert(my.index());
        return intersect;
    }
    
    @Override
    public void erase(Object input)
    {
        if(count((T) input) == 1)
            for(int i = 0; i < _size; ++i)
                if(_set[i] == (T) input)
                {
                    --_size;
                    while(i < _size)
                    {
                        _set[i] = _set[i + 1];
                        ++i;
                    }
                }
    }
    
    @Override
    public void clear()
    {
        _size = 0;
        _capacity = 10;
        _set = (T[]) new Object [_capacity];
    }
    
    @Override
    public GTUIterator<T> find(Object input)
    {
        int i;
        for(i = 0; i < _size; ++i)
            if(_set[i] == (T) input)
                break;
        return new GTUIterator<>((T[])_set, i, _size);
    }
    
    @Override
    public int count(Object input)
    {
        int counter = 0;
        for(int i = 0; i < _size && counter == 0; ++i)
            if(_set[i] == (T) input)
                ++counter;
        return counter;
    }

    @Override
    public GTUIterator<T> begin()
    {
        return new GTUIterator<>((T[])_set, 0, _size);
    }

    @Override
    public GTUIterator<T> end()
    {
        return new GTUIterator<>((T[])_set, _size, _size);
    }
    @Override
    public String toString()
    {
        String value = "{\n";
        for(GTUIterator<T> search = begin(); search.hasNext(); search.next())
        {
            value += "\t";
            value += search;
            value += "\n";
        }
        value += "}";
       return value; 
    }
}